arrow-core-data/src/main/kotlin/arrow/core/Option.kt

```diff
@@ -614,10 +1043,317 @@ fun <T> Iterable<T>.lastOrNone(predicate: (T) -> Boolean): Option<T> = this.last
 fun <T> Iterable<T>.elementAtOrNone(index: Int): Option<T> = this.elementAtOrNull(index).toOption()
 
 fun <A, B> Option<Either<A, B>>.select(f: OptionOf<(A) -> B>): Option<B> =
-  flatMap { it.fold({ l -> Option.just(l).ap(f) }, { r -> Option.just(r) }) }
+  branch(f.fix(), Some(::identity))
+
+fun <A, B, C> Option<Either<A, B>>.branch(fa: Option<(A) -> C>, fb: Option<(B) -> C>): Option<C> =
+  flatMap { it.fold(
+    { a -> Some(a).ap(fa) },
+    { b -> Some(b).ap(fb) }
+  )}
+
+private fun Option<Boolean>.selector(): Option<Either<Unit, Unit>> =
+  map { bool -> if (bool) Either.right(Unit) else Either.left(Unit) }
+
+fun <A> Option<Boolean>.whenS(x: Option<() -> Unit>): Option<Unit> =
+  selector().select(x.map { f -> { _: Unit -> f() } })
+
+fun <A> Option<Boolean>.ifS(fl: Option<A>, fr: Option<A>): Option<A> =
+  selector().branch(fl.map { { _: Unit -> it } }, fr.map { { _: Unit -> it } })
+
+fun Option<Boolean>.orS(f: Option<Boolean>): Option<Boolean> =
+  ifS(Option.just(true), f)
+
+fun Option<Boolean>.andS(f: Option<Boolean>): Option<Boolean> =
+  ifS(f, Option.just(false))
+
+fun <A> Option<A>.combineAll(MA: Monoid<A>): A = MA.run {
+  foldLeft(empty()) { acc, a -> acc.combine(a) } }
+
+inline fun <A> Option<A>.ensure(error: () -> Unit, predicate: (A) -> Boolean): Option<A> =
+  when (this) {
+    is Some -> if (predicate(t)) this else Option.raiseError(error())
+    is None -> this
+  }
+
+/**
+ * Returns an Option containing all elements that are instances of specified type parameter R.
+ */
+inline fun <reified B> Option<*>.filterIsInstance(): Option<B> {
+  val f: (Any?) -> B? = { it as? B }
+  return this.filterMap(f)
+}
+
+inline fun <A> Option<A>.handleError(f: (Unit) -> A): Option<A> =
+  handleErrorWith { Option.just(f(Unit)) }
+
+inline fun <A> Option<A>.handleErrorWith(f: (Unit) -> Option<A>): Option<A> =
+  if (isEmpty()) f(Unit) else this
+
+inline fun <reified B> Option<*>.traverseFilterIsInstance(): List<Option<B>> =
+  filterA { a -> listOf(a is B) }.map { it.map { a -> a as B } }
+
+inline fun <E, reified B> Option<*>.traverseFilterIsInstanceEither(): Either<E, Option<B>> =
+  filterAEither { a -> Right(a is B) }.map { it.map { a -> a as B } }
```

Comment:
How does this function work? 🤔 This will result in a `ClassCastException` when `a` is not `B`.